dwh/indicator_lib.py

```diff
@@ -0,0 +1,149 @@
+"""Set of functions to work with Spark DataFrames containing FHIR resources.
+
+See test_spark.ipynb for real examples of how to create/use these functions.
+"""
+
+from typing import List
+from datetime import datetime
+import pandas
+from pyspark.sql import DataFrame
+import pyspark.sql.functions as F
+import pyspark.sql.types as T
+
+
+def flatten_obs(obs: DataFrame) -> DataFrame:
+  """Creates a flat version of Observation FHIR resources.
+
+  Args:
+    obs: A collection of Observation FHIR resources.
+  Returns:
+    A DataFrame with three columns where `coding` arrays are flattened.
+  """
+  return obs.select(
+      obs.subject.patientId.alias('patientId'),
+      obs.effective.dateTime.alias('dateTime'),
+      obs.value,
+      F.explode(obs.code.coding).alias('coding'))
+
+
+def aggregate_all_codes_per_patient(
+    obs: DataFrame,
+    codes: List[str]=None,
+    start_date: str=None,
+    end_date: str=None) -> DataFrame:
+  """ For each patientId, generates aggregate values for all their observations.
+
+  Args:
+    obs: A collection of Observation FHIR resources.
+    codes: A list of codes for which aggregates are generated or None to
+      indicate all codes.
+    start_date: The first date before which Observations are dropped.
+    end_date: The last date after which Observations are dropped.
+  Returns:
+    A DataFrame with one row for each patient and several aggregate columns for
+    each Observation code.
+  """
+  flat_obs = flatten_obs(obs)
+  start_obs = flat_obs
+  if start_date:
+    start_obs = flat_obs.filter(flat_obs.dateTime > start_date)
+  date_obs = start_obs
+  if end_date:
+    date_obs = start_obs.filter(start_obs.dateTime < end_date)
+  return date_obs.groupBy([
+      flat_obs.patientId,
+  ]).pivot('coding.code', values=codes).agg(
+      F.count('*').alias('num_obs'),
+      F.min(flat_obs.value.quantity.value).alias('min_value'),
+      F.max(flat_obs.value.quantity.value).alias('max_value'),
+      F.min(flat_obs.dateTime).alias('min_date'),
+      F.max(flat_obs.dateTime).alias('max_date')
+  )
+
+
+def join_patients_agg_obs(
+    patients: DataFrame,
+    agg_obs: DataFrame,
+    base_patient_url: str):
+  """Joins a collection of Patient FHIR resources with an aggregated obs set.
+
+  Args:
+    patients: A collection of Patient FHIR resources.
+    agg_obs: Aggregated observations from `aggregate_all_codes_per_patient()`.
+  Returns:
+    Same `agg_obs` with corresponding patient information joined.
+  """
+  flat_patients = patients.select(
+      patients.id, patients.birthDate, patients.gender).withColumn(
+      'actual_id', F.regexp_replace('id', base_patient_url, ''))
+  return flat_patients.join(
+      agg_obs, flat_patients.actual_id == agg_obs.patientId)
+
+
+def find_age_band(birth_date: str) -> str:
+  """Given the birth date, finds the age_band for PEPFAR disaggregation."""
+  birth = datetime.strptime(birth_date, '%Y-%m-%d')
+  age = int((datetime.today() - birth).days / 365.25)
```

Comment:
would be nice to pass in a parameter instead of using `datetime.today()` there are situations where you want to retrospectively generate a report/analytics e.g as of 2019, how many patients aged >19 yo   had VLvl suppression